src/openms/source/FORMAT/MzQCFile.cpp

```diff
@@ -0,0 +1,234 @@
+// --------------------------------------------------------------------------
+
+#include <OpenMS/FORMAT/MzQCFile.h>
+
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/algorithm/copy.hpp>
+#include <nlohmann/json.hpp>
+
+using namespace std;
+
+// TODO fix all the shadowed "const_iterator qpsit"
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wshadow"
+
+namespace OpenMS
+{ 
+  void MzQCFile::store(const String & inputFileName,
+                       const String & outputFileName,
+                       MSExperiment exp,
+                       ControlledVocabulary cv,
+                       const String & contactName,
+                       const String & contactAddress,
+                       const String & description,
+                       const String & label) const
+  {
+    using json = nlohmann::ordered_json;
+    json out;
+    // required: creationDate, version
+    DateTime currentTime = DateTime::now();
+    out["MzQC"]["creationDate"] = currentTime.toString();
+    out["MzQC"]["version"] = "1.0.0";
+
+    // optional: contactName, contactAddress, description
+    if (!contactName.empty()) 
+    {
+      out["MzQC"]["contactName"] = contactName;
+    }
+    if (!contactAddress.empty()) 
+    {
+      out["MzQC"]["contactAddress"] = contactAddress;
+    }
+    if (!description.empty()) 
+    {
+      out["MzQC"]["description"] = description;
+    }
+
+    out["MzQC"]["runQualities"]["metadata"]["label"] = label;
+    out["MzQC"]["runQualities"]["metadata"]["inputFiles"] = {
+        {
+          {"location", String(QFileInfo(QString::fromStdString(inputFileName)).absoluteFilePath())},
+          {"name", String(QFileInfo(QString::fromStdString(inputFileName)).baseName())},
+          {"fileFormat",
+          {
+            {"accession", "MS:10000584"},
+            {"name", "mzML format"}
+          }
+          }
+        },
+          "fileProperties",
+          {
+            {
+              {"accession", "MS:1000747"},
+              {"name", "completion time"},
+              {"value", String(exp.getDateTime().getDate() + " " + exp.getDateTime().getTime())}
+            },
+            {
+              {"accession", "MS:1000569"},
+              {"name", "SHA-1"},
+              {"value", String(FileHandler::computeFileHash(inputFileName))}
+            },
+            {
+              {"accession", "MS:1000031"},
+              {"name", "instrument model"},
+              {"value", String(exp.getInstrument().getName())}
+            }
+          }
+    };
+
+    out["MzQC"]["runQualities"]["metadata"]["analysisSoftware"] = { // todo
+        {
+          {"accession", "MS:1009001" }, // create new qc-cv for QCCalculator: MS:1009001 quality control metrics generating software
+          {"name", "QCCalculator"},
+          {"version", "2.6.0"},
+          {"uri", "https://www.openms.de"}
+        }
+    };
+
+    out["MzQC"]["runQualities"]["qualityMetrics"] = {};
```

Comment:
because it is an ordered json that contains the input order of elements. The idea was to build the scaffold of the json first and then add all the quality metrics afterwards.